.buildkite/scripts/serverless/create_deploy_tag/create_fix_tag.sh

```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+set -euo pipefail
+
+git cat-file -e "$BUILDKITE_COMMIT"
```

Comment:
This is just checking that `BUILDKITE_COMMIT` is a valid object, which the `git tag` command below does as well... I _think_ this should correctly validate if `BUILDKITE_COMMIT` is a commit on `BUILDKITE_BRANCH`:

```
git merge-base --is-ancestor `git merge-base HEAD main` "$BUILDKITE_COMMIT"
```

The above command expects `BUILDKITE_BRANCH` to be the currently checked out branch. If that's not the case, guess `HEAD` can be replaced by `$BUILDKITE_BRANCH`

Reply:
Can you help me understand the nested `git merge-base HEAD main` part?  I'm not sure why the merge-base for main is needed.

Reply:
I pushed c28648b08a2b70c584bb6eba33e05497d07daa91

Reply:
Sure :) What I'm trying to do is verify if the commit we're about to tag actually exists on the specified `deploy-fix@` branch. Technically you could by accident paste a commit SHA for a commit that's actually not on the branch and the `git tag` command have no way of validating this.

So I imagine a git tree that looks like this:

```
A - B - C - E (main)
        \- B1 - B2 (deploy-fix@xxx)
```

Here `E` is the head of `main` and `B2` is the head of the `deploy-fix@` branch, which branched of `main` at point `B`. The task is to get a list of commits containing only `B1` and `B2` and then check that `BUILDKITE_COMMIT` is one of these.

The command `git merge-base HEAD main` works off the assumption that `deploy-fix@xxx` is the currently checked out branch, which is what `HEAD` is referring to. If you wasn't sure which branch was checked out, you could re-write it as `git merge-base deploy-fix@xxx main` to be sure. This command will return `B` - the common ansestor between the two branches.

Next I feed in that result (`B`) to the command `git merge-base --is-ancestor B "$BUILDKITE_COMMIT"`, and this checks if `B` is an ancestor to `BUILDKITE_COMMIT`... hmm now that I write that I can see that it technically would also be true of `BUILDKITE_COMMIT` was `E` - so I guess that part needs to be re-written 🤔 Let me just think about that...

Reply:
Ok, I re-wrote this to not be a one-liner, to make it a bit easier to read:

```bash
# Step 1: Check if the commit is in the specific named branch
if git merge-base --is-ancestor $BUILDKITE_COMMIT $BUILDKITE_BRANCH; then
    echo "Commit $BUILDKITE_COMMIT is part of the $BUILDKITE_BRANCH branch"
    
    # Step 2: Check if the commit is also part of any other branches
    # This command lists all branches containing the commit and counts them
    branches_containing_commit=$(git branch --contains $BUILDKITE_COMMIT | wc -l)
    
    # If the commit is in more than one branch, exit with non-zero
    if [[ $branches_containing_commit -gt 1 ]]; then
        echo "Error: Commit $BUILDKITE_COMMIT is part of multiple branches."
        exit 1
    else
        echo "Commit $BUILDKITE_COMMIT is uniquely part of the $BUILDKITE_BRANCH branch."
    fi
else
    echo "Commit $BUILDKITE_COMMIT is not part of the $BUILDKITE_BRANCH branch."
    exit 1
fi
```

Feel free to re-write

Reply:
Thank you, that's helpful.  I pushed 470be0371bb68fd33750d0204f6c067e3f7d2641